foundationdb/src/tuple.rs

```diff
@@ -0,0 +1,364 @@
+use byteorder::{self, ByteOrder};
+use std;
+use std::io::Write;
+
+#[derive(Debug, Clone, Copy)]
+pub enum TupleError {
+    //
+    EOF,
+    InvalidType,
+    InvalidData,
+}
+
+type Result<T> = std::result::Result<T, TupleError>;
+
+trait Single: Sized {
+    fn encode<W: Write>(&self, _w: W) -> std::io::Result<()>;
+    fn encode_to_vec(&self) -> std::io::Result<Vec<u8>> {
+        let mut v = Vec::new();
+        self.encode(&mut v)?;
+        Ok(v)
+    }
+
+    fn decode(buf: &[u8]) -> Result<(Self, usize)>;
+    fn decode_full(buf: &[u8]) -> Result<Self> {
+        let (val, offset) = Self::decode(buf)?;
+        if offset != buf.len() {
+            return Err(TupleError::InvalidData);
+        }
+        Ok(val)
+    }
+}
+
+impl Single for bool {
+    fn encode<W: Write>(&self, mut w: W) -> std::io::Result<()> {
+        if *self {
+            w.write_all(&[0x27])
```

Comment:
We should define an enum for these codes, you can copy that from my initial tuple impl...

Reply:
See the SingleType from that with it’s associated to_code impl.